sql/api/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -117,65 +130,144 @@ object JavaTypeInference {
-      val classTV = JavaTypeUtils.getTypeArguments(c, classOf[Object]).asScala.toMap ++
-        typeVariables
-      // Note that the fields are ordered by name.
-      val fields = properties.map { property =>
-        val readMethod = property.getReadMethod
-        val encoder = encoderFor(readMethod.getGenericReturnType, seenTypeSet + c, classTV)
-        // The existence of `javax.annotation.Nonnull`, means this field is not nullable.
-        val hasNonNull = readMethod.isAnnotationPresent(classOf[Nonnull])
-        EncoderField(
-          property.getName,
-          encoder,
-          encoder.nullable && !hasNonNull,
-          Metadata.empty,
-          Option(readMethod.getName),
-          Option(property.getWriteMethod).map(_.getName))
+      // if the properties is empty and this is not a top level enclosing class, then we
+      // should not consider class as bean, as otherwise it will be treated as empty schema
+      // and loose the data on deser.
+      if (properties.isEmpty && seenTypeSet.nonEmpty) {
+        if (classOf[KryoSerializable].isAssignableFrom(c)) {
+          Encoders.kryo(c).asInstanceOf[AgnosticEncoder[_]]
+        } else if (classOf[java.io.Serializable].isAssignableFrom(c)) {
+          Encoders.javaSerialization(c).asInstanceOf[AgnosticEncoder[_]]
+        } else {
+          throw ExecutionErrors.cannotFindEncoderForTypeError(t.getTypeName)
+        }
+      } else {
+        // add type variables from inheritance hierarchy of the class
+        val parentClassesTypeMap =
+          JavaTypeUtils.getTypeArguments(c, classOf[Object]).asScala.toMap
+        val classTV = parentClassesTypeMap ++ typeVariables
+        // Note that the fields are ordered by name.
+        val fields = properties.map { property =>
+          val readMethod = property.getReadMethod
+          val methodReturnType = readMethod.getGenericReturnType
+          val encoder = encoderFor(methodReturnType, seenTypeSet + c, classTV)
+          // The existence of `javax.annotation.Nonnull`, means this field is not nullable.
+          val hasNonNull = readMethod.isAnnotationPresent(classOf[Nonnull])
+          EncoderField(
+            property.getName,
+            encoder,
+            encoder.nullable && !hasNonNull,
+            Metadata.empty,
+            Option(readMethod.getName),
+            Option(property.getWriteMethod).map(_.getName))
+        }
+        // implies it cannot be assumed a BeanClass.
+        // Check if its super class or interface could be represented by an Encoder
+
+        JavaBeanEncoder(ClassTag(c), fields.toImmutableArraySeq)
       }
-      JavaBeanEncoder(ClassTag(c), fields.toImmutableArraySeq)
 
-    case _ =>
-      throw ExecutionErrors.cannotFindEncoderForTypeError(t.toString)
+    case t =>
+      throw ExecutionErrors.cannotFindEncoderForTypeError(t.getTypeName)
   }
 
+  private def getAllSuperClasses(typee: Type): Array[Class[_]] = Option(typee)
```

Comment:
This sort of begs for a queue and a loop. I am reasonable sure that is more readable...

Reply:
will flatten the recursion

Reply:
@hvanhovell Done. Kindly check.